### ISSUE TITLE
Using APM with SOAP

### ISSUE THREAD

**Issue by dhtek** [elastic/apm-agent-dotnet#637]
I would like to use APM in a SOAP web service project. I succeeded to use ElasticApmModule based on the sample "AspNetFullFrameworkSampleApp" from the repository. 

Nevertheless, I'm not really happy because the SoapAction is added in the header of the request and it's not to use to build the transaction name. So, for a same path, we aggregate all the actions.

Ex: 
/Search.asmx 
-> Action GetCustomers
-> Action GetFamillies

Solution : 
Allow to extend the ElasticApmModule with a custom code to override the transaction name ?

**Comment by SergeyKleyman** [comment]
Thank you for opening an issue, @dhtek

Regarding the future improvements:

> Allow to extend the ElasticApmModule with a custom code to override the transaction name ?

Could you please give a example/general idea how you would like to make `ElasticApmModule` aware of the custom code to override the transaction name?

Regarding what you can do with the Agent's current implementation (version 1.2) - you can use [Agent's public API](https://www.elastic.co/guide/en/apm/agent/dotnet/current/public-api.html#api-current-transaction) to set [`Name`](https://github.com/elastic/apm-agent-dotnet/blob/master/src/Elastic.Apm/Api/IExecutionSegment.cs#L43) property at any moment during the processing the transaction by your code. The Agent will use `Name` value at the end of the transaction.

**Comment by dhtek** [comment]
I would prefer not to add some codes related to APM in our projects by referencing directly the public API. APM is working well using the httpmodule and allow me to have a solution "plug & play" with all our projects.

Currently, for our POC, I referenced directly APM projects in our solution and replace these lines 

[Start transaction](https://github.com/elastic/apm-agent-dotnet/blob/b82c07e55dd1c1dfbfa662c82fc1cf3aa8658b1a/src/Elastic.Apm.AspNetFullFramework/ElasticApmModule.cs#L113) and [Start transaction](https://github.com/elastic/apm-agent-dotnet/blob/b82c07e55dd1c1dfbfa662c82fc1cf3aa8658b1a/src/Elastic.Apm.AspNetFullFramework/ElasticApmModule.cs#L120)

with 

`_currentTransaction = Agent.Instance.Tracer.StartTransaction($"{httpRequest.HttpMethod} {httpRequest.Path} {httpRequest.Headers.SoapAction()}".Trim(), ApiConstants.TypeRequest, distributedTracingData);`

It's was a quick solution for our POC and just adding this code `httpRequest.Headers.SoapAction()` resolved our issue (with Trim()).

What do you think of creating a new module for SOAP inheriting from ElasticApmModule and overriding the transaction name ?

**Comment by SergeyKleyman** [comment]
> What do you think of creating a new module for SOAP inheriting from ElasticApmModule and overriding the transaction name ?

Creating new module just to cover this particular use case seems like an overkill - I do wonder why not to make it the default behavior (using `SOAPAction` HTTP request header if present) of `ElasticApmModule` itself. A possible issue I see is that `SOAPAction` value can be quite long ([for example](https://docs.microsoft.com/en-us/openspecs/sql_server_protocols/ms-ssnws/bb3d9ea3-6f43-4c72-afc2-c873b304dcde) `http://schemas.microsoft.com/sqlserver/2004/SOAPsqlbatch`) @dhtek Is it as long in your case? If so does the resulting transaction name look okay to you in Kibana?

@gregkalapos WDYT?

**Comment by gregkalapos** [comment]
> > What do you think of creating a new module for SOAP inheriting from ElasticApmModule and overriding the transaction name ?
> 
> Creating new module just to cover this particular use case seems like an overkill - I do wonder why not to make it the default behavior (using `SOAPAction` HTTP request header if present) of `ElasticApmModule` itself. A possible issue I see is that `SOAPAction` value can be quite long ([for example](https://docs.microsoft.com/en-us/openspecs/sql_server_protocols/ms-ssnws/bb3d9ea3-6f43-4c72-afc2-c873b304dcde) `http://schemas.microsoft.com/sqlserver/2004/SOAPsqlbatch`) @dhtek Is it as long in your case? If so does the resulting transaction name look okay to you in Kibana?
> 
> @gregkalapos WDYT?

I think that'd be a good approach. 

@dhtek Is this the "old" `.asmx` SOAP project type that you use? 

The `httpRequest.Headers.SoapAction()` is probably a method in a dll that is present in that project type (or something you implemented), which is not present in `Elastic.Apm.AspNetFullFramework`, but if this is a known header name, as @SergeyKleyman says we can easily search for that header manually.

Does SOAP always add this HTTP header? It'd be nice to know if there are cases when reading the header does not work and if there are, what those cases are. From what I understand reading the SOAP message would give us the true action name, but reading the xml would be probably an overkill. 

In any case, I think looking for the header name is a good first approach.

**Comment by dhtek** [comment]
@SergeyKleyman Yes the "Soapaction" in the header is too long to display in kibana, I splitted the path to take only the last word corresponding to the action.

@gregkalapos  Yes it's the old asmx... 😒

This is the code that I used:

```
public static string SoapAction(this NameValueCollection httpHeaders) =>
httpHeaders
.Get("Soapaction")?
.Split('/')?
.Last()?
.Replace("\\", "")
.Replace("\"", "");
```

I removed the backslash and the quotation mark from the soapaction. The result:

![image](https://user-images.githubusercontent.com/2010406/70521372-6800e480-1b3f-11ea-8ccc-b2589f5dfb28.png)

Using the trim() at the end allows me to use the module with soap and webapi without condition

**Comment by dhtek** [comment]
Any news about that ?

**Comment by SergeyKleyman** [comment]
> Any news about that ?

Not yet - we do welcome community PRs if you would like to give it a try.

**Comment by dhtek** [comment]
After investigation, I noticed that soap action is not more present in http header for soap 1.2. It is only present in the body of the request. Currently, I just implemented the code for soap 1.1 because I think it's the most common used and the default protocol is soap 1.1 when using the soap client.
